transformer_engine/common/layer_norm/norms.cpp

```diff
@@ -0,0 +1,350 @@
+/*************************************************************************
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See LICENSE for license information.
+ ************************************************************************/
+
+/* #include <transformer_engine/layer_norm.h> */
+
+#include "norms.h"
+
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <numeric>
+
+/*
+
+Supported Type combinations:
+
+input    compute   weights   output
+=======================================
+fp32     fp32      fp32      fp32
+fp16     fp32      fp16      fp16
+bf16     fp32      bf16      bf16
+fp32     fp32      fp16      fp16
+fp32     fp32      bf16      bf16
+bf16     fp32      bf16      fp8
+
+Remarks:
+Output type = Weight type
+Compute always in FP32
+
+*/
+
+namespace transformer_engine {
+
+// Create registries and provide runtime versions of config hash functions.
+FwdTunedRegistry LN_FWD_TUNED_FUNCS;
+BwdTunedRegistry LN_BWD_TUNED_FUNCS;
+FwdGeneralRegistry LN_FWD_GENERAL_FUNCS;
+BwdGeneralRegistry LN_BWD_GENERAL_FUNCS;
+
+FwdTunedRegistry RMS_FWD_TUNED_FUNCS;
+BwdTunedRegistry RMS_BWD_TUNED_FUNCS;
+FwdGeneralRegistry RMS_FWD_GENERAL_FUNCS;
+BwdGeneralRegistry RMS_BWD_GENERAL_FUNCS;
+
+uint32_t get_type_id(DType dtype) {
+  if (dtype == DType::kFloat16) {
+    return TypeId<fp16>::Value;
+  } else if (dtype == DType::kBFloat16) {
+    return TypeId<bf16>::Value;
+  } else if (dtype == DType::kFloat32) {
+    return TypeId<fp32>::Value;
+  } else if (dtype == DType::kFloat8E4M3) {
+    return TypeId<fp8e4m3>::Value;
+  } else {
+    NVTE_ERROR("Type not supported.");
+  }
+}
+
+uint64_t get_key(DType wtype, DType itype, DType otype, DType ctype, uint64_t hidden_size) {
+  uint64_t type_key = get_type_id(wtype) | (get_type_id(itype) << 2) | (get_type_id(otype) << 4) |
+                      (get_type_id(ctype) << 6);
+  uint64_t launcher_key = (type_key << 32) | hidden_size;
+  return launcher_key;
+}
+
+template <NVTE_NORM_TYPE NormEnum>
+FwdFunction& get_fwd_launcher(DType wtype, DType itype, DType otype, DType ctype,
+                              const FwdParams& params) {
+  if constexpr (!IF_TE_FWD_NORMS<NormEnum>()) NVTE_ERROR("Unexpected NVTE_NORM_TYPE!");
```

Comment:
We can catch these errors at compile-time with a `static_assert`:
```suggestion
  static_assert(IF_TE_FWD_NORMS<NormEnum>(), "Unexpected NVTE_NORM_TYPE");
```
I see this `if constexpr (...) { NVTE_ERROR(...) }` pattern used in several other places.